### PR TITLE
Rocm Dockerfile for AMD GPU support 

### LINKED ISSUE #13609: Adding a Dockerfile to support Lightning on ROCm and AMD GPUs
## 🚀 Feature

 The proposed feature would be the addition of a Dockerfile to the directory lightning/dockers that builds Lightning for ROCm.

### Motivation

 The motivation would be ease of use for people wanting to work with Lightning who use ROCm and AMD GPUs.

### Pitch

 The simple addition of a Dockerfile that builds Lightning for ROCm. This Dockerfile would be building upon the latest ROCm Pytorch Dockerfile.

### Alternatives

 None currently

### Additional context

 None currently

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @carmocca @akihironitta @borda @justusschock @kaushikb11 @awaelchli @rohitgr7

### PR DESCRIPTION
**Pull request by sgschwindAMD**
## What does this PR do?

Adds a Dockerfile to build Lightning for ROCm 

Closes #13609.

### Does your PR introduce any breaking changes?

 None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @pruthvistony

### CODE REVIEW

File: dockers/for_rocm/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    pip install jsonargparse && \
+    pip install -U jsonargparse[signatures]
+
+#for flash 
+RUN python -m pip install urllib3
+
+#for bolts and transformers
+RUN pip install protobuf==3.9.2
+
+RUN \
```
Comment (justusschock) [inline dockers/for_rocm/Dockerfile:22]: Can we actually keep this similar to https://github.com/Lightning-AI/lightning/blob/master/dockers/base-cuda/Dockerfile ? So that they are virtually the same except for rocm vs cuda?

**Review by justusschock — COMMENTED** [review]
Also when we do add the dockerfile here, we at least need a CI building that image :)

File: dockers/README.md [hunk 2]
```diff
@@ -86,3 +86,19 @@ Inspiration comes from https://u.group/thinking/how-to-put-jupyter-notebooks-in-
...
    - replace the `hostname` by `localhost`
+
+## Build docker image for Lightning on ROCm 
+
+To build the docker image for ROCm follow these steps: 
+
+```bash
+git clone <git-repository>
+#cd into the newly cloned repository
+docker build -f dockers/for_rocm/Dockerfile .
```
Comment (Borda) [inline dockers/README.md:97]:
```suggestion
docker build -f dockers/rocm/Dockerfile .
```
this seems rather similar

File: dockers/for_rocm/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    pip install click && \
+    pip install deepdiff && \
+    pip install fastapi && \
+    pip install lightning-cloud && \
+    pip install -U Jinja2 && \
+    pip install starsessions && \ 
+    pip uninstall fsspec && \ 
+    pip install fsspec==2022.1.0 && \ 
+    pip install jsonargparse && \
+    pip install -U jsonargparse[signatures]
```
Comment (Borda) [inline dockers/for_rocm/Dockerfile:14]: why not install it as a single cmd so that eventually collisions are resolved?

**Comment by sgschwindAMD** [comment]
Thanks for the feed back, I'll work on these changes as well as changing the dockerfile it builds FROM from rocm/pytorch:latest to rocm/pytorch:latest-release.

File: dockers/README.md [hunk 4]
```diff
@@ -14,10 +14,10 @@ or with specific arguments
 ```bash
 git clone <git-repository>
 docker image build \
-    -t pytorch-lightning:base-cuda-py3.9-pt1.10 \
-    -f dockers/base-cuda/Dockerfile \
+    -t pytorch-lightning:base-cuda-py3.9-pt1.11 \
+    -f dockers/base-rocm/Dockerfile \
```
Comment (akihironitta) [inline dockers/README.md:18]: I would keep `base-cuda` since no users are expected to use `base-rocm` until PL supports ROCm.
```suggestion
    -f dockers/base-cuda/Dockerfile \
```

File: dockers/release-rocm/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,40 @@
+ARG PYTHON_VERSION=3.9
```
Comment (akihironitta) [inline dockers/release-rocm/Dockerfile:1]: I think we shouldn't add this for now because
1. If I see it correctly, this dockerfile isn't specific to rocm at all, so we can simply reuse `dockers/release/Dockerfile`, and
2. we won't release docker image for ROCm anyway until PL officially supports it.

File: dockers/for_rocm/Dockerfile (hunk 1 above)
Comment (sgschwindAMD) [reply dockers/for_rocm/Dockerfile]: > Also when we do add the dockerfile here, we at least need a CI building that image :)

As to the CI building, we're looking if we can provide a machine for that but I am unsure when I can get a machine for this. Do you have any github action resources free to build the docker image?

Comment (sgschwindAMD) [reply dockers/for_rocm/Dockerfile]: > Can we actually keep this similar to https://github.com/Lightning-AI/lightning/blob/master/dockers/base-cuda/Dockerfile ? So that they are virtually the same except for rocm vs cuda?

We have updated the base-rocm file to be as similar to the base-cuda file as possible.

File: dockers/README.md (hunk 2 above)
Comment (sgschwindAMD) [reply dockers/README.md]: Taken care of

File: dockers/for_rocm/Dockerfile (hunk 3 above)
Comment (sgschwindAMD) [reply dockers/for_rocm/Dockerfile]: taken care of in the newest commit

File: dockers/for_rocm/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+#for flash 
+RUN python -m pip install urllib3
```
[the remark this answers is not in the document]
Comment (sgschwindAMD) [reply dockers/for_rocm/Dockerfile]: Updated in latest commit, please review the new base-rocm file

File: dockers/README.md (hunk 4 above)
Comment (sgschwindAMD) [reply dockers/README.md]: Corrected, sorry about that

File: dockers/release-rocm/Dockerfile (hunk 5 above)
Comment (sgschwindAMD) [reply dockers/release-rocm/Dockerfile]: We removed the release-rocm folder and just shifted the Dockerfile from there into the base-rocm folder with a different file name

**Comment by sgschwindAMD** [comment]
Please review the restructured files in the base-rocm folder. We made significant changes to the structure of the code to mirror the base-cuda Dockerfile more closely

File: dockers/for_rocm/Dockerfile (hunk 1 above)
Comment (akihironitta) [reply dockers/for_rocm/Dockerfile]: > Do you have any github action resources free to build the docker image?

@sgschwindAMD I tried pushing a commit to your branch, but it seems like you don't allow us to push any commits. Please cherry-pick this commit: https://github.com/Lightning-AI/lightning/commit/756692335d8a8d21c45fc58738dc31d246d35f59

File: dockers/README.md [hunk 7]
```diff
@@ -86,3 +86,28 @@ Inspiration comes from https://u.group/thinking/how-to-put-jupyter-notebooks-in-
...
    - replace the `hostname` by `localhost`
+
+## Build docker image for Lightning on ROCm
+
+To build the docker image for ROCm follow these steps:
+
+```bash
+git clone <git-repository>
+docker image build \
+    -t pytorch_lightning:base-rocm-py3.9-torch1.12.0 \
```
Comment (akihironitta) [inline dockers/README.md:97]: Similarly to #13831, do you think we should add rocm_whl_version to the tag?
```suggestion
    -t pytorch_lightning:base-rocm-py3.9-torch1.12.0-rocm5.1.1 \
```

File: dockers/README.md [hunk 8]
```diff
@@ -86,3 +86,28 @@ Inspiration comes from https://u.group/thinking/how-to-put-jupyter-notebooks-in-
...
+    -f dockers/base-rocm/Dockerfile \
+    --build-arg PYTHON_VERSION=3.9 \
+    --build-arg PYTORCH_VERSION=1.12.0 \
+    .
+
+docker image build \
+    -t pytorch_lightning:lightning-rocm-py3.9-torch1.12.0 \
```
Comment (akihironitta) [inline dockers/README.md:104]: Same as above.
```suggestion
    -t pytorch_lightning:lightning-rocm-py3.9-torch1.12.0-rocm5.1.1 \
```

**Review by akihironitta — CHANGES_REQUESTED** [review]
@sgschwindAMD @pruthvistony Would you mind including 756692335d8a8d21c45fc58738dc31d246d35f59 in this branch as commented in https://github.com/Lightning-AI/lightning/pull/13610#discussion_r937245766? Without this commit, we can't verify the change made in this PR in CI.

**Comment by pruthvistony** [comment]
> @sgschwindAMD @pruthvistony Would you mind including [7566923](https://github.com/Lightning-AI/lightning/commit/756692335d8a8d21c45fc58738dc31d246d35f59) in this branch as commented in [#13610 (comment)](https://github.com/Lightning-AI/lightning/pull/13610#discussion_r937245766)? Without this commit, we can't verify the change made in this PR in CI.
> 
> Also, following comments in #13609, let's not merge this PR for now. Even if we don't merge this, developers are still able to build docker images from the Dockerfile and work on ROCm integration.

I have cherry-picked [7566923](https://github.com/Lightning-AI/lightning/commit/756692335d8a8d21c45fc58738dc31d246d35f59) into the PR. Thanks for the change.

**Comment by pruthvistony** [comment]
> > @sgschwindAMD @pruthvistony Would you mind including [7566923](https://github.com/Lightning-AI/lightning/commit/756692335d8a8d21c45fc58738dc31d246d35f59) in this branch as commented in [#13610 (comment)](https://github.com/Lightning-AI/lightning/pull/13610#discussion_r937245766)? Without this commit, we can't verify the change made in this PR in CI.
> > Also, following comments in #13609, let's not merge this PR for now. Even if we don't merge this, developers are still able to build docker images from the Dockerfile and work on ROCm integration.
> 
> I have cherry-picked [7566923](https://github.com/Lightning-AI/lightning/commit/756692335d8a8d21c45fc58738dc31d246d35f59) into the PR. Thanks for the change.

@akihironitta ,
Thanks for the github actions changes, I see the build-rocm job executing.

**Comment by pruthvistony** [comment]
@akihironitta @Borda ,
Please let us know if PR can be merged?

**Comment by justusschock** [comment]
@pruthvistony It cannot be merged before we are able to verify PL with AMD GPUs (see https://github.com/Lightning-AI/lightning/issues/13609#issuecomment-1204851165 for reference).

**Comment by Borda** [comment]
@pruthvistony sorry for delay, let me check it this week :otter:

**Comment by carmocca** [comment]
Closing as we don't have access to amd hardware at the moment

**Comment by IncubatorShokuhou** [comment]
@sgschwindAMD Hi, can you please tell me if [ROCmSoftwarePlatform:ROCm_support](https://github.com/ROCmSoftwarePlatform/pytorch-lightning/tree/ROCm_support) is working properly at this moment?

**Comment by pruthvistony** [comment]
> @sgschwindAMD Hi, can you please tell me if [ROCmSoftwarePlatform:ROCm_support](https://github.com/ROCmSoftwarePlatform/pytorch-lightning/tree/ROCm_support) is working properly at this moment?

@IncubatorShokuhou ,
Can you please let me know what do you mean by working properly on ROCm? 
When the PR was raised it was working properly.
Now obviously the fork [ROCmSoftwarePlatform:ROCm_support](https://github.com/ROCmSoftwarePlatform/pytorch-lightning/tree/ROCm_support) is quiet old and wasnt updated.

We want to add support for ROCm (AMD GPUs) in Lightning-AI, so please let me know if you have any AMD hardware access. We can work on this and update the PR and provide support.

cc @carmocca

**Comment by carmocca** [comment]
@pruthvistony We would need to setup a CI job that runs on AMD hardware to be confident about its support. That's what we do for all other accelerators. Would you be able to help with that?

cc @Borda

**Comment by IncubatorShokuhou** [comment]
> > @sgschwindAMD Hi, can you please tell me if [ROCmSoftwarePlatform:ROCm_support](https://github.com/ROCmSoftwarePlatform/pytorch-lightning/tree/ROCm_support) is working properly at this moment?
> 
> @IncubatorShokuhou , Can you please let me know what do you mean by working properly on ROCm? When the PR was raised it was working properly. Now obviously the fork [ROCmSoftwarePlatform:ROCm_support](https://github.com/ROCmSoftwarePlatform/pytorch-lightning/tree/ROCm_support) is quiet old and wasnt updated.
> 
> We want to add support for ROCm (AMD GPUs) in Lightning-AI, so please let me know if you have any AMD hardware access. We can work on this and update the PR and provide support.
> 
> cc @carmocca

I am currently getting ready to utilize lightning on a ROCm machine, but I haven't gained access to the machine yet. I sincerely appreciate the opportunity to receive your valuable assistance, and I am truly grateful for your help.

**Comment by pruthvistony** [comment]
@carmocca @IncubatorShokuhou ,
Regarding getting CI machines, I need to check internally with AMD infrastructure team/budget and will get back soon.

**Comment by Borda** [comment]
> Regarding getting CI machines, I need to check internally with AMD infrastructure team/budget and will get back soon.

That would be great! Let me know if I can help...
